src/v/cloud_storage_clients/abs_client.cc

```diff
@@ -0,0 +1,584 @@
+    // DELETE /{container-id}/{blob-id} HTTP/1.1
+    // Host: {storage-account-id}.blob.core.windows.net
+    // x-ms-date:{req-datetime in RFC9110}
+    // x-ms-version:"2021-08-06"
+    // Authorization:{signature}
+    const auto target = fmt::format("/{}/{}", name(), key().string());
+    const boost::beast::string_view host{_ap().data(), _ap().length()};
+
+    http::client::request_header header{};
+    header.method(boost::beast::http::verb::delete_);
+    header.target(target);
+    header.insert(boost::beast::http::field::host, host);
+    header.insert("x-ms-delete-snapshots", "include");
+
+    auto error_code = _apply_credentials->add_auth(header);
+    if (error_code) {
+        return error_code;
+    }
+
+    return header;
+}
+
+result<http::client::request_header>
+abs_request_creator::make_list_blobs_request(
+  const bucket_name& name,
+  std::optional<object_key> prefix,
+  [[maybe_unused]] std::optional<object_key> start_after,
+  std::optional<size_t> max_keys) {
+    // GET /{container-id}?restype=container&comp=list&prefix={prefix}...
+    // ...&max_results{max_keys}
+    // HTTP/1.1 Host: {storage-account-id}.blob.core.windows.net
+    // x-ms-date:{req-datetime in RFC9110}
+    // x-ms-version:"2021-08-06"
+    // Authorization:{signature}
+    auto target = fmt::format("/{}?restype=container&comp=list", name());
+    if (prefix) {
+        target += fmt::format("&prefix={}", prefix.value()().string());
+    }
+
+    if (max_keys) {
+        target += fmt::format("&max_results={}", max_keys.value());
+    }
+
+    const boost::beast::string_view host{_ap().data(), _ap().length()};
+
+    http::client::request_header header{};
+    header.method(boost::beast::http::verb::get);
+    header.target(target);
+    header.insert(boost::beast::http::field::host, host);
+
+    auto error_code = _apply_credentials->add_auth(header);
+    if (error_code) {
+        return error_code;
+    }
+
+    return header;
+}
+
+abs_client::abs_client(
+  const abs_configuration& conf,
+  ss::lw_shared_ptr<const cloud_roles::apply_credentials> apply_credentials)
+  : _requestor(conf, std::move(apply_credentials))
+  , _client(conf)
+  , _probe(conf._probe) {}
+
+abs_client::abs_client(
+  const abs_configuration& conf,
+  const ss::abort_source& as,
+  ss::lw_shared_ptr<const cloud_roles::apply_credentials> apply_credentials)
+  : _requestor(conf, std::move(apply_credentials))
+  , _client(conf, &as, conf._probe, conf.max_idle_time)
+  , _probe(conf._probe) {}
+
+ss::future<> abs_client::stop() { return _client.stop(); }
+
+void abs_client::shutdown() { _client.shutdown(); }
+
+template<typename T>
+ss::future<result<T, error_outcome>> abs_client::send_request(
+  ss::future<T> request_future,
+  const bucket_name& bucket,
+  const object_key& key) {
+    using namespace boost::beast::http;
+
+    auto outcome = error_outcome::retry;
+
+    try {
+        co_return co_await std::move(request_future);
```

Comment:
nothing wrong with this pattern, but it seems like it would be more common to call this method "handle_request_exception(std::exception_ptr)" rather than "send_request" and have the caller wait on the future. nothing wrong with it, just curious if it makes something else simpler.

Reply:
I see your point. I did something similar to what you're suggesting first and then switched to this when implementing the requests (like you noticed in https://github.com/redpanda-data/redpanda/pull/8086#discussion_r1068649628).

Reply:
yeh, it makes things pretty nice and clean!